disperser/controller/metadata/batch_metadata_manager.go

```diff
@@ -0,0 +1,199 @@
+		ctx:                  ctx,
+		logger:               logger,
+		ethClient:            ethClient,
+		metadata:             atomic.Pointer[BatchMetadata]{},
+		indexedChainState:    indexedChainState,
+		registryCoordinator:  registryCoordinator,
+		referenceBlockOffset: referenceBlockOffset,
+		updatePeriod:         updatePeriod,
+	}
+	manager.alive.Store(true)
+
+	// Make sure we have valid metadata before the constructor returns.
+	err = manager.updateMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update initial metadata: %w", err)
+	}
+
+	go manager.updateLoop()
+
+	return manager, nil
+}
+
+// GetMetadata returns the most recent batch metadata. This method is thread safe.
+func (m *batchMetadataManager) GetMetadata() *BatchMetadata {
+	return m.metadata.Load()
+}
+
+// Close releases resources associated with this manager.
+func (m *batchMetadataManager) Close() {
+	m.alive.Store(false)
+}
+
+// Fetch the next reference block number (RBN) to use.
+func (m *batchMetadataManager) getNextReferenceBlockNumber() (uint64, error) {
+	// Get the latest block header to determine the current reference block number.
+	latestHeader, err := m.ethClient.HeaderByNumber(m.ctx, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get latest block header: %w", err)
+	}
+	latestBlockNumber := latestHeader.Number.Uint64()
+
+	if latestBlockNumber < m.referenceBlockOffset {
+		return 0, fmt.Errorf("latest block number is less than RBN offset: %d < %d",
+			latestBlockNumber, m.referenceBlockOffset)
+	}
+
+	return latestBlockNumber - m.referenceBlockOffset, nil
+}
+
+// get a list of all quorums that are registered for a particular reference block number.
+func (m *batchMetadataManager) getQuorums(referenceBlockNumber uint64) ([]core.QuorumID, error) {
+
+	// Quorums are assigned starting at 0, and then sequentially without gaps. If we
+	// know the number of quorums, we can generate a list of quorum IDs.
+
+	quorumCount, err := m.registryCoordinator.QuorumCount(&bind.CallOpts{
+		Context:     m.ctx,
+		BlockNumber: new(big.Int).SetUint64(referenceBlockNumber),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get quorum count: %w", err)
+	}
+
+	quorums := make([]core.QuorumID, quorumCount)
+	for i := uint8(0); i < quorumCount; i++ {
+		quorums[i] = i
+	}
+
+	return quorums, nil
+}
+
+// updateMetadata fetches the latest batch metadata from the blockchain and updates m.operatorState.
+// This method is called periodically to ensure that metadata reflects a recent(ish) reference block.
+func (m *batchMetadataManager) updateMetadata() error {
+	referenceBlockNumber, err := m.getNextReferenceBlockNumber()
+	if err != nil {
+		return fmt.Errorf("failed to get next reference block number: %w", err)
+	}
+
+	previousMetadata := m.metadata.Load()
+	if previousMetadata != nil && referenceBlockNumber < previousMetadata.referenceBlockNumber {
+		// Only update if the new RBN is greater than the most recent one.
```

Comment:
This would probably indicate a bug if it actually happened, right? Would it make sense to log?